syntax = "proto3";

package chat;

// Message definitions
message ChatRoom {
  string id = 1;
  string name = 2;
}

message Message {
  string id = 1;
  string room_id = 2;
  string author_id = 3;
  string content = 4;
  int64 timestamp = 5;
}

message CreateRoomRequest {
  string name = 1;
}

message CreateRoomResponse {
  ChatRoom room = 1;
}

message ListRoomsRequest {}

message ListRoomsResponse {
  repeated ChatRoom rooms = 1;
}

message JoinRoomRequest {
  string room_id = 1;
  string author_id = 2;
}

message JoinRoomResponse {
  Message message = 1;
}

message SendMessageRequest {
  string room_id = 1;
  string author_id = 2;
  string content = 3;
}

message SendMessageResponse {
  Message message = 1;
}

message GetHistoryRequest {
  string room_id = 1;
}

message GetHistoryResponse {
  repeated Message messages = 1;
}

// Service definition
service ChatService {
  // Create a new chat room
  rpc CreateRoom(CreateRoomRequest) returns (CreateRoomResponse) {}
  
  // List all chat rooms
  rpc ListRooms(ListRoomsRequest) returns (ListRoomsResponse) {}
  
  // Join a room and receive streamed messages
  rpc JoinRoom(JoinRoomRequest) returns (stream JoinRoomResponse) {}
  
  // Send and receive messages in bidirectional stream
  rpc SendMessage(stream SendMessageRequest) returns (stream SendMessageResponse) {}
  
  // Get message history for a room
  rpc GetHistory(GetHistoryRequest) returns (GetHistoryResponse) {}
}